src/ua/main.cpp

```diff
@@ -529,6 +538,63 @@ void printEnvironmentInfo() {
   }
 }
 
+// There is currently the possibility of a race condition if a chunk
+// upload timed-out.  It's possible that a second upload succeeds,
+// has the chunk marked as "complete" and then the first request makes
+// its way through the queue and marks the chunk as pending again.
+// Since we are just about to close the file, we'll check to see if any
+// chunks are marked as pending, and if so, we'll retry them.
+void check_for_complete_chunks(vector<File> &files) {
+  for (int currCheckNum=0; currCheckNum < NUM_CHUNK_CHECKS; ++currCheckNum){
+    map<string, JSON> fileDescriptions;
+    while (!chunksFinished.empty()) {
+      Chunk *c = chunksFinished.consume();
+
+      // Cache file descriptions so we only have to do once per file,
+      // not once per chunk.
+      if (fileDescriptions.find(c->fileID) == fileDescriptions.end())
+          fileDescriptions[c->fileID] = fileDescribe(c->fileID);
+
+      if (!is_chunk_complete(c, fileDescriptions[c->fileID])) {
```

Comment:
I also found another hiccup which should have been obvious to me, but became apparent when running the test you suggested.  Once the chunk is uploaded, the actual data in the chunk is deleted.  This obviously makes sense because otherwise by the end of the upload you'd have all the data you uploaded present in memory!  So rather than simply push the chunks to be retried into the upload queue, we should push them into the read queue so that the data is reloaded, optionally compressed, and then uploaded.  This actually simplifies the code I think.  

Anyway, with these changes, I then ran the test you suggested:
I uploaded a 72 MB file in 5 MB chunks and had each chunk pretend to fail 50% of the time.  I allowed up to 30 retry attempts (only needed 3-4 of course).  I then uploaded the same file using the current version of ua.  I downloaded the two files and ran md5sum to verify that they have the same hash.  They do.  

I ran this test with ua both with and without a compression step.  Both checked out good.